### ISSUE TITLE
[Inductor] Constant folding for dynamic shape `full` node before pattern matching

### ISSUE THREAD

**Issue by imzhuhl** [pytorch/pytorch#128387]
### 🚀 The feature, motivation and pitch

Hi, I'm having a problem with constant folding in dynamic shape. 
Here is the code for a simple reproduction of this situation:
```python
import torch

def foo(x: torch.Tensor):
    batch, length = x.shape
    mask = torch.ones((1, length), device=x.device, dtype=x.dtype)
    mask = 1 - mask
    tmp = x + mask
    return tmp * 2

x = torch.randn((2, 4))
with torch.no_grad():
    torch.compile(foo, dynamic=True)(x)
```
Here is the graph after `constant_fold_uniform_value` in `joint_graph_passes`:
```python
#  arg0_1: "Sym(s0)", arg1_1: "Sym(s1)", arg2_1: "f32[s0, s1]"
graph():
    %arg0_1 : [num_users=0] = placeholder[target=arg0_1]
    %arg1_1 : [num_users=1] = placeholder[target=arg1_1]
    %arg2_1 : [num_users=1] = placeholder[target=arg2_1]
    %full : [num_users=1] = call_function[target=torch.ops.aten.full.default](args = ([1, %arg1_1], 1), kwargs = {dtype: torch.float32, layout: torch.strided, device: cpu, pin_memory: False})
    %sub : [num_users=1] = call_function[target=torch.ops.aten.sub.Tensor](args = (1, %full), kwargs = {})
    %add : [num_users=1] = call_function[target=torch.ops.aten.add.Tensor](args = (%arg2_1, %sub), kwargs = {})
    %mul : [num_users=1] = call_function[target=torch.ops.aten.mul.Tensor](args = (%add, 2), kwargs = {})
    return (mul,)
```
If close dynamic shape, the graph is:
```python
graph():
    %arg0_1 : [num_users=1] = placeholder[target=arg0_1]
    %full : [num_users=0] = call_function[target=torch.ops.aten.full.default](args = ([1, 4], 1), kwargs = {dtype: torch.float32, layout: torch.strided, device: cpu, pin_memory: False})
    %full_default : [num_users=0] = call_function[target=torch.ops.aten.full.default](args = ([1, 4], 0.0), kwargs = {dtype: torch.float32, layout: torch.strided, device: cpu, pin_memory: False})
    %mul : [num_users=1] = call_function[target=torch.ops.aten.mul.Tensor](args = (%arg0_1, 2), kwargs = {})
    return (mul,)
```
Replacing sub node as full_default, and folding add node.
Such folding cannot be done under dynamic shape,  but logically, we can fold mask tensor. By the way, I think we also can remove two full nodes in final graph (not the focus of this issue).

This improvement will help a lot of sdpa calls to flash attention(no mask) instead of memory efficient attention(with mask), thereby significantly improving performance.

I have seen this PR https://github.com/pytorch/pytorch/pull/118411 (for unbacked symint) and would like to ask for your opinions @ezyang @eellison. Is it possible to do constant folding with hint size(specialize full and save the value).

cc @ezyang @bdhirsh @anijain2305 @chauhang

**Comment by eellison** [comment]
@imzhuhl 

I think we can extend `ConstantFoldUniformValue` to:

1. Use substitute size for ops which have dynamic shape inputs (could be hint but not sure it matters) 
2. Keep track of which values have been produced by dynamic shapes, only allow their use in pointwise ops/views
3. When we have a constant value that is produced by symint, insert the symint full instead of constant fold.


Would you have any interest in submitting this pr ?

**Comment by imzhuhl** [comment]
> @imzhuhl
> 
> I think we can extend `ConstantFoldUniformValue` to:
> 
> 1. Use substitute size for ops which have dynamic shape inputs (could be hint but not sure it matters)
> 2. Keep track of which values have been produced by dynamic shapes, only allow their use in pointwise ops/views
> 3. When we have a constant value that is produced by symint, insert the symint full instead of constant fold.
> 
> Would you have any interest in submitting this pr ?

Okay, I can try this
